Planner_add command implementation

WendyBaiYunwei opened the pull request:
Tests will be implemented through another PR

WendyBaiYunwei committed:
DegreePlanner: implement add command

WendyBaiYunwei committed:
Planner: amend planner add command

WendyBaiYunwei committed:
Merge branch 'master' into planner

WendyBaiYunwei committed:
planner: resovle merge conflicts

WendyBaiYunwei committed:
planner: delete yunwei's original planner implementation

WendyBaiYunwei committed:
planner: implement planner_add 1

WendyBaiYunwei committed:
Planner: implement planner_add command 2

WendyBaiYunwei committed:
planner: implement planner_add command

WendyBaiYunwei committed:
Planner: amend test

WendyBaiYunwei committed:
planner: amend style

WendyBaiYunwei committed:
Planner: resolve merge conflicts

WendyBaiYunwei committed:
Merge branch 'master' into planner_add

WendyBaiYunwei committed:
planner_add: remove extra lines

WendyBaiYunwei committed:
Merge branch 'master' into planner_add

WendyBaiYunwei committed:
planner_add: amend errors

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
planner_add: fix planner_add parser

WendyBaiYunwei committed:
planner_add: fix command implementation

WendyBaiYunwei committed:
planner_add: fix command implementation

WendyBaiYunwei committed:
planner_add: fix other minor issues

WendyBaiYunwei committed:
planner_add: fix style issues

lycjackie reviewed (changes requested):
Some comments based on user testing.

1.  After a successful add. The display panel on the right is not refreshing.
2. `planner_add` help message did not reflect ` ` between `y/` and `s/`
3. `planner_add code/CS2040C y/1 s/2` lead to an runtime error

i.e. `planner_add code/CS1231 y/1 s/1` is okay but 
`planner_add code/CS2040C y/1 s/2` is not

WendyBaiYunwei committed:
planner_add: fix command implementation

WendyBaiYunwei committed:
planner_add: remove an unused line

WendyBaiYunwei committed:
planner_add: remove an extra line

WendyBaiYunwei committed:
planner_add: fix style

lycjackie commented on src/main/java/seedu/address/logic/parser/CliSyntax.java:
```diff
@@ -9,7 +9,8 @@
+    public static final Prefix PREFIX_SEMESTER = new Prefix("s/");
     public static final Prefix PREFIX_TAG = new Prefix("tag/");
     public static final Prefix PREFIX_COREQUISITE = new Prefix("coreq/");
-
```
Unnecessary deletion of line

WendyBaiYunwei replied:
there will be a checkstyle error if i delete the line o.o

WendyBaiYunwei replied:
don't delete*

lycjackie commented on src/main/java/seedu/address/model/DegreePlannerList.java:
```diff
@@ -60,6 +60,14 @@ public boolean hasDegreePlanner(DegreePlanner degreePlanner) {
+    /**
+     * Returns true if a module with the same identity as {@code degreePlanner} exists in the particular degreePlanner.
+     */
+    public boolean hasDegreePlannerModule(DegreePlanner plannerModules) {
```
Could have a better variable name instead of `plannerModules`

`plannerModules` normally suggest a `list` of something.

lycjackie commented on src/main/java/seedu/address/model/DegreePlannerList.java:
```diff
@@ -68,6 +76,10 @@ public void addDegreePlanner(DegreePlanner p) {
         degreePlanners.add(p);
     }
 
+    public void addDegreePlannerModule(DegreePlanner p) {
```
Likewise, single variable name is not recommended.

lycjackie commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -168,6 +168,8 @@
      */
     boolean hasDegreePlanner(DegreePlanner degreePlanner);
 
+    boolean hasDegreePlannerModule(DegreePlanner plannerModules);
```
As mention previously

lycjackie commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -180,6 +182,11 @@
+     * Adds the given module to the degree planner.
+     * {@code plannerModule} must not already exist in the address book.
+     */
+    void addDegreePlannerModule(DegreePlanner plannerModules);
```
As mention previously

lycjackie commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -295,6 +295,12 @@ public boolean hasDegreePlanner(DegreePlanner planner) {
     }
 
+    @Override
+    public boolean hasDegreePlannerModule(DegreePlanner plannerModules) {
```
As mention previously

lycjackie commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -3,14 +3,16 @@
+import seedu.address.model.module.Code;
 import seedu.address.model.planner.exceptions.DegreePlannerNotFoundException;
 import seedu.address.model.planner.exceptions.DuplicateDegreePlannerException;
-
```
Unnecessary deletion of line.

WendyBaiYunwei replied:
line resumed

lycjackie commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -49,6 +50,36 @@ public void add(DegreePlanner toAdd) {
+     */
+    public int location(DegreePlanner toCheck) {
+        requireNonNull(toCheck);
+        int location = 0;
```
if `location = 0` from the start, even if the checks fail. 

This will still return `location = 0` which can be a valid index.

@chyeo You also have this mistake. It was my bad that I did not spot it during the review.

WendyBaiYunwei replied:
location variable is required to get initialized. Correct me if I'm wrong. I don't think this part will cause problem because we already checked that year and sem are valid number. thus the returned index will be valid as well.

lycjackie replied:
Yes, you're right that location variable is required for this.

However, if you feel that there is no logic bug in this. Then I'm okay with this. 

@Creastery did mention in our chat on why using location is bad too btw. 

But do bear in mind that if a bug is found regarding this. You might still need to fix
this again...

lycjackie commented on src/main/java/seedu/address/model/planner/Year.java:
```diff
@@ -10,14 +10,15 @@
     public static final String YEAR_VALIDATION_REGEX =
             "[1-4]{1}";
 
+
```
Unnecessary line added

lycjackie commented on src/main/java/seedu/address/storage/Storage.java:
```diff
@@ -62,3 +62,4 @@
 
 
 }
+
```
Unnecessary line added

Try not to make any modification to a file which this PR is not related.

lycjackie commented on src/main/java/seedu/address/storage/StorageManager.java:
```diff
@@ -164,3 +164,4 @@ public void saveRequirementCategoryList(ReadOnlyAddressBook requirementCategoryL
         addressBookStorage.saveRequirementCategoryList(requirementCategoryList, filePath);
     }
 }
+
```
Unnecessary line added

lycjackie commented on src/test/java/seedu/address/logic/commands/AddCommandTest.java:
```diff
@@ -237,6 +237,10 @@ public ReadOnlyDegreePlannerList getDegreePlannerList() {
         }
 
+        @Override
+        public boolean hasDegreePlannerModule(DegreePlanner plannerModules) {
```
As mention previously, do take note of all your variable name.

lycjackie commented on src/test/java/seedu/address/logic/commands/AddCommandTest.java:
```diff
@@ -237,6 +237,10 @@ public ReadOnlyDegreePlannerList getDegreePlannerList() {
 
+        @Override
+        public boolean hasDegreePlannerModule(DegreePlanner plannerModules) {
+            return false;
```
Since this is a test class

Is either you implement it correctly or you leave it empty for the moment.

By returning a static value, this is a bad impression that your code is running as expected
if there is any test cases that use this method.

WendyBaiYunwei replied:
I added a toDo comment and am planning to implement the test afterwards. is it ok? according to tutors one pr doesn't need to contain everything. PR can be small o.o

lycjackie replied:
Yes okay. Do remember to update your TODO

lycjackie commented:
After the latest build, here some stuff I found.

1. Able to add non-existence module into the planner. (is this suppose to be like this?)
2. The display panel is still not updating after the command. 
- Look at `edit` command for example.

WendyBaiYunwei committed:
planner_add: check if modules to add exist
planner_add command can add in any non-existing modules.
let's check whether the modules are inside the application.
the modules can only be added when they exist in the application.

WendyBaiYunwei committed:
planner_add: change prefix format for year and sem

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -28,7 +33,6 @@
     private final ObservableList<DegreePlanner> internalList = FXCollections.observableArrayList();
     private final ObservableList<DegreePlanner> internalUnmodifiableList =
             FXCollections.unmodifiableObservableList(internalList);
-
```
there will be a checkstyle error if I don't delete the line

WendyBaiYunwei replied:
line resumed

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: add comment

WendyBaiYunwei committed:
planner_add: fix style

lycjackie commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+        if (model.hasDegreePlannerModules(toAdd)) {
+            throw new CommandException(MESSAGE_DUPLICATE_MODULE);
+        }
+        if (!model.existingPlannerModules(toAdd, model)) {
```
You might want to reference the below code to improve on this method.
https://github.com/CS2113-AY1819S2-T09-1/main/blob/7d55c1a32c117cea8ca5548763c6d358b183cfb7/src/main/java/seedu/address/logic/commands/RequirementAddCommand.java#L63-L65

This does not really make sense that an object `model` take in itself as a parameter...

lycjackie commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+package seedu.address.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.logic.commands.RequirementAddCommand.MESSAGE_MODULE_DOES_NOT_EXIST;
```
Why are you using `RequirementAddCommand` message for your own planner?

You might want to write your own message.

lycjackie commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -136,4 +172,47 @@ private boolean degreePlannersAreUnique(List<DegreePlanner> degreePlanners) {
+    /**
+     * checks if the modules are contained inside the entire degree planner
+     */
+    public boolean containsModules(DegreePlanner toCheck) {
```
Might want to consider how `UniqueRequirementCategory.java` implements this...

https://github.com/CS2113-AY1819S2-T09-1/main/blob/7d55c1a32c117cea8ca5548763c6d358b183cfb7/src/main/java/seedu/address/model/requirement/UniqueRequirementCategoryList.java#L49-L52

This allow our code base to be consistent.

lycjackie commented:
I have clone down your current branch and tried running the command.

I seem to be encountering errors by inputting

`planner_add code/CS1231 year/1 sem/1`

Do try to get this fix too..

WendyBaiYunwei committed:
planner_add: minor enhancements

WendyBaiYunwei committed:
Planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: resolve errors

some codes went missing / wrong when resolving merge conflicts.

let's fix them.

WendyBaiYunwei commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -334,15 +334,36 @@ public boolean hasDegreePlanner(DegreePlanner degreePlanner) {
     public void addDegreePlanner(DegreePlanner degreePlanner) {
         degreePlanners.add(degreePlanner);
     }
-
```
Line deleted because it caused checkstyle error

WendyBaiYunwei replied:
line resumed

WendyBaiYunwei commented on src/test/java/seedu/address/logic/commands/AddCommandTest.java:
```diff
@@ -77,7 +77,6 @@ public void equals() {
 
         // same object -> returns true
         assertTrue(addAliceCommand.equals(addAliceCommand));
-
```
there are a few trailing white space checkstyle issues. perhaps it is my configuration problem. But can we ignore the extra(?) deletion of the line if it does not affect the code quality?

WendyBaiYunwei replied:
line resumed

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: change semester validation

Letm3through commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -334,7 +334,13 @@ public boolean hasDegreePlanner(DegreePlanner degreePlanner) {
         requireNonNull(degreePlanner);
         return degreePlanners.contains(degreePlanner);
     }
-
```
May I know what is the purpose of removing this line?
I think you should not remove this line.

WendyBaiYunwei replied:
line resumed

Letm3through commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -334,7 +334,13 @@ public boolean hasDegreePlanner(DegreePlanner degreePlanner) {
+    public boolean hasDegreePlannerModules(DegreePlanner plannerModules) {
+        requireNonNull(plannerModules);
+        return degreePlanners.containsModules(plannerModules);
+    }
```
Please add a blank line between like the rest

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -343,6 +349,22 @@ public void addDegreePlanner(DegreePlanner degreePlanner) {
+        degreePlanners.addModules(p);
+        indicateModified();
+    }
+    /**
```
Please add the blank line between as stated above

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -343,6 +349,22 @@ public void addDegreePlanner(DegreePlanner degreePlanner) {
+        requireNonNull(plannerCode);
+        return modules.asUnmodifiableObservableList().stream().anyMatch((module) ->
+                module.getCode().equals(plannerCode));
+    }
```
Please add the blank line between as stated above

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -191,6 +193,11 @@
 
+    /**
+     * Adds the given module to the degree planner.
+     * {@code plannerModule} must not already exist in the address book.
```
Do you mean `plannerModules`?

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/planner/Semester.java:
```diff
@@ -10,13 +10,13 @@
 
     public static final String MESSAGE_SEMESTER_CONSTRAINTS =
-            "Semester should only be given in a number range from 1 to 4, S (e.g. 1) format and in valid form.";
+            "Semester should be either 1 or 2, in format sem/1 or sem/2.";
```
Try to edit `Semester should be either 1 or 2, in format sem/1 or sem/2.` to make it consistent with other constraints related messages.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/planner/Year.java:
```diff
@@ -10,7 +10,7 @@
 
     public static final String MESSAGE_YEAR_CONSTRAINTS =
-            "Year in an university should only be given in Y (e.g. 1) format and in valid form.";
+            "Year should be in y/x format, where x is an integer greater than 0 and smaller than 5.";
```
As we discuss, we will be using `year/` instead of `y/`. 
Please amend this part. 
Also, I think it will be better to use `between` instead of saying `greater ....` or `less` instead of `smaller`

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -49,6 +53,36 @@ public void add(DegreePlanner toAdd) {
+    }
+
+    /**
+     * Adds the new modules to the particular degree planner with matching year and semester.
```
Perhaps, you could be more clear by rephrasing the sentence instead of just stating `new modules`

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -136,4 +170,23 @@ private boolean degreePlannersAreUnique(List<DegreePlanner> degreePlanners) {
         return true;
     }
+    /**
+     * Returns true if the modules to add are contained inside the entire degree planner.
```
Do you mean any one of the modules? Might want to state it more clearly.
Also as mentioned by Jackie, you may want to change the name of the method to be consistent.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -49,6 +53,36 @@ public void add(DegreePlanner toAdd) {
         internalList.add(toAdd);
     }
 
+    /**
```
I think you shouldn't implement the `planner_add` method this way.
You should implement in the manner where `planner_add` edit degree planner by setting `new` degree planner based on modules requested to be added to the degree planner. 
This way `planner_delete` will able to utilize the same methods too.
In addition, you would not need the method `location` for it.
If you are not sure, please refer to how others(e.g. `requirement_add`) are implemented.
Please try to make changes.

WendyBaiYunwei replied:
Thank you for all the feedback. I have implemented planner_delete locally, and it takes in different parameter (a set of codes). Due to different format, it will not reuse planner_add implementation. (as planner_delete does not specify the year and sem for convenience reason)

WendyBaiYunwei replied:
the code quality is assessed individually, thus the quality of my code will not affect the team. Still, I welcome feedback because I also want to improve. But just mentioning the assessment in case you do not know.

WendyBaiYunwei replied:
The location method I will just leave it at least for now because of a couple of reasons. I need to implement co-requisite and write for test. and I do not have much time left. Besides, I do not want my code to be identical as CH's code because I cannot think of his method myself.
Please also note that consistency among group members are not assessed as well.

Letm3through replied:
> Besides, I do not want my code to be identical as CH's code because I cannot think of his method myself.

This method you are using is just identical to the version of Requirement Add at #82  before the bug fix in #112:
https://github.com/chyeo/main/blob/576f6049c23c52863ce82a5fc5c9d6e0f4d46d30/src/main/java/seedu/address/model/requirement/UniqueRequirementCategoryList.java#L53-L63
So I am not sure why you are mentioning the point. 
I am still not sure why you have to use locate the degree planner in this manner.
I thought we have discussed about this matter previously in the group.
May I know what is actually changed in this method after the discussion so that I have clearer idea about this?

WendyBaiYunwei replied:
I mean I could think of the location method, but stream was not something I could think of. I have modified the implementation.

Letm3through commented on src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java:
```diff
@@ -49,6 +53,36 @@ public void add(DegreePlanner toAdd) {
+    /**
+     * Adds the new modules to the particular degree planner with matching year and semester.
+     */
+    public void addModules(DegreePlanner toAdd) {
```
Refer to the abovementioned comment at line at https://github.com/CS2113-AY1819S2-T09-1/main/blob/f8a31042aecb88072c9a9c220ab4193f5321229e/src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java#L59

Letm3through commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -293,6 +293,12 @@ public boolean hasDegreePlanner(DegreePlanner planner) {
     }
 
+    @Override
+    public boolean hasDegreePlannerModules(DegreePlanner plannerModules) {
```
Is this necessary to be implemented here? I think you can directly implement this in `PlannerAddCommand`.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -301,6 +307,11 @@ public boolean hasDegreePlanner(DegreePlanner planner) {
         versionedAddressBook.addDegreePlanner(degreePlanner);
     }
 
+    @Override
```
Same comment as mentioned above.

Letm3through commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -375,6 +386,14 @@ public void setSelectedRequirementCategory(RequirementCategory requirementCatego
         selectedRequirementCategory.setValue(requirementCategory);
     }
 
+    /**
```
Same comment as mentioned above.

Letm3through commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -179,6 +179,8 @@
      */
     boolean hasDegreePlanner(DegreePlanner degreePlanner);
 
+    boolean hasDegreePlannerModules(DegreePlanner plannerModules);
```
Refer to the ModelManager comment at line https://github.com/CS2113-AY1819S2-T09-1/main/blob/f8a31042aecb88072c9a9c220ab4193f5321229e/src/main/java/seedu/address/model/ModelManager.java#L297

Letm3through commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -191,6 +193,11 @@
+     * Adds the given module to the degree planner.
+     * {@code plannerModule} must not already exist in the address book.
+     */
+    void addDegreePlannerModules(DegreePlanner plannerModules);
```
Refer to the comment at line https://github.com/CS2113-AY1819S2-T09-1/main/blob/f8a31042aecb88072c9a9c220ab4193f5321229e/src/main/java/seedu/address/model/ModelManager.java#L311

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,61 @@
+
+    public static final String COMMAND_WORD = "planner_add";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a module to the degree planner. "
```
I realised you are trying to add modules. If then, please correct this `COMMAND_WORD`

WendyBaiYunwei replied:
planner means the modules to add are related to the degree planner. It's meant to differentiate between normal adding and adding to planner. I thought we already agreed on the name. if i make it "module_add", it can be misleading as ppl don't know whether it's for planner?

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,61 @@
+            + PREFIX_YEAR + "YEAR "
+            + PREFIX_SEMESTER + "SEMESTER";
+
+    public static final String MESSAGE_SUCCESS = "New module added: %1$s";
```
What if you are adding more than one module?
You may want to include the case too.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,61 @@
+            + PREFIX_SEMESTER + "SEMESTER";
+
+    public static final String MESSAGE_SUCCESS = "New module added: %1$s";
+    public static final String MESSAGE_DUPLICATE_MODULE = "This module already exists in the degree planner";
```
What are more than one module existing in the degree planner?

WendyBaiYunwei replied:
fixed

WendyBaiYunwei committed:
planner_add: change message

WendyBaiYunwei committed:
planner_add: modify command messages

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+
+    public static final String COMMAND_WORD = "planner_add";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds modules to degree planner. "
```
You may want to edit it to `Adds modules to a degree planner.` since you are adding on the particular degree planner

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+            + PREFIX_YEAR + "YEAR "
+            + PREFIX_SEMESTER + "SEMESTER";
+
+    public static final String MESSAGE_SUCCESS = "New modules added to degree planner: %1$s";
```
What if you are adding just one module?
You may want to include the case too.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+
+    public static final String MESSAGE_SUCCESS = "New modules added to degree planner: %1$s";
+    public static final String MESSAGE_DUPLICATE_MODULE = "Some of the modules already exist in degree planner";
+    public static final String MESSAGE_MODULE_DOES_NOT_EXIST = "Some of the module codes do"
```
You used `modules` and here you are using `module codes`.
You may want to edit to be consistent.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+            + PREFIX_SEMESTER + "SEMESTER";
+
+    public static final String MESSAGE_SUCCESS = "New modules added to degree planner: %1$s";
+    public static final String MESSAGE_DUPLICATE_MODULE = "Some of the modules already exist in degree planner";
```
What if only one module exists? You may want to include that.

WendyBaiYunwei replied:
fixed

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
```
May want to better implement this part. Refer to the comment at line https://github.com/CS2113-AY1819S2-T09-1/main/blob/f8a31042aecb88072c9a9c220ab4193f5321229e/src/main/java/seedu/address/model/planner/UniqueDegreePlannerList.java#L59

WendyBaiYunwei replied:
fixed

Letm3through reviewed (changes requested):
Please consider making following changes to the way `Planner_add` is implemented.
In addition, you should have your own `PlannerAddCommandTest`.
Also, please remove irrelevant methods u added in `AddCommandTest` if there is after making changes. 
Thanks you.

WendyBaiYunwei committed:
planner_add: fix errors

WendyBaiYunwei committed:
planner_add: remove extra test

WendyBaiYunwei commented:
> Please consider making following changes to the way `Planner_add` is implemented.
> In addition, you should have your own `PlannerAddCommandTest`.
> Also, please remove irrelevant methods u added in `AddCommandTest` if there is after making changes.
> Thanks you.

Thank you so much for your advice. the extra tests have been removed and new tests will be submitted through another PR.

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei commented:
Please note that after I remove the extra tests, appveyor check fails. Please let me know ASAP if you need me to resume the tests or make any other changes. Thank you!

Letm3through commented:
@WendyBaiYunwei What I meant above was to `remove irrelevant methods u added in AddCommandTest if there is after making changes` which means if you are removing methods (for e.g. addDegreePlannerModules) in the process of changing implementation of `planner_add`, please remember to remove this method in the `AddCommandTest`. I mentioned this part because I believe that you can have better implementation of `planner_add` and during the process of modification, you may remove some methods you added in like `addDegreePlannerModules`.
Sorry for making this unclear.

WendyBaiYunwei committed:
planner_add: resume tests

lycjackie commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,62 @@
+            + "Parameters: "
+            + PREFIX_CODE + "CODE "
+            + PREFIX_YEAR + "YEAR "
+            + PREFIX_SEMESTER + "SEMESTER";
```
Perhaps you might want to consider adding examples on your command usage?

This would help the users to have a better idea on what to input

Pretty useful for tester too when we can just copy and paste the examples.

Creastery commented:
There are already two reviewers, which have requested changes previously.
You should fix them before asking them for a review.

WendyBaiYunwei committed:
planner_add: amend implementation

WendyBaiYunwei committed:
planner_add: fix comments

WendyBaiYunwei committed:
planner_add: modify validating module implementation
existing methods of identifying duplicate and invalid modules are not consistent
let's fix it and make them consistent

WendyBaiYunwei committed:
planner_add: remove extra part

WendyBaiYunwei committed:
planner_add: fix errors

WendyBaiYunwei committed:
planner_add: change implementation
planner_add command takes in degree planner has parameter
let's change the degree planner parameter to year, semester and codes

WendyBaiYunwei committed:
planner_add: fix test

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: fix comment

Letm3through commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,85 @@
+            throw new CommandException(MESSAGE_DUPLICATE_MODULE);
+        }
+
+        if (codesToAdd.stream().anyMatch(code -> !model.existingModuleListModules(code))) {
```
Is there need to implement method for this and not use the pre-existing one in the `moduleList`?

WendyBaiYunwei replied:
fixed

Letm3through reviewed (changes requested):
Please look through the codes one more time before adding `status.ToReview` so that we don't have to review multiple times on similar issues.
Furthmore, please resolve conversation when it is resolved so that it is clear which issue you have resolved.

WendyBaiYunwei committed:
planner_add: fix semester message

WendyBaiYunwei committed:
planner_add: remove existingModuleListModule

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: add in usage example

WendyBaiYunwei committed:
planner_add: resume hasdegreeplanner

WendyBaiYunwei committed:
planner_add: resume extra deletion

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
planner_add: resume extra deletion (+1 squashed commits)

Squashed commits:

[0eb310b4] planner_add: resume extra deletion (+1 squashed commits)

Squashed commits:

[16028a8d] planner_add: remove extra deletion

WendyBaiYunwei committed:
planner_add: fix messages

lycjackie commented:
The `Logic` part of the code generally looks okay. 

As mention in person previously,
I will request a review from @Creastery on the `Model` part.

I guess should be ready to merge soon after he takes a look on the `Model`

Do remember to add in your unit tests for your functionality.

WendyBaiYunwei committed:
planner_add: remove extra parts

WendyBaiYunwei committed:
planner_add: amend comments

WendyBaiYunwei committed:
planner_add: amend comments (+1 squashed commits)

Squashed commits:

[5e56b2b2] planner_add: amend comments

Creastery commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -271,6 +276,21 @@ public void addDegreePlanner(DegreePlanner degreePlanner) {
+        for (DegreePlanner existingCode : streamSet) {
+            existingCodes.addAll(existingCode.getCodes());
+        }
+        return existingCodes.stream().anyMatch((existingCode) -> existingCode.equals(plannerCode));
```
The above code suffers from performance issues.
You can simply do:
```java
return getDegreePlannerList().stream().map(DegreePlanner::getCodes)
        .anyMatch(codes -> codes.contains(plannerCode));
```

Creastery commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -271,6 +276,21 @@ public void addDegreePlanner(DegreePlanner degreePlanner) {
+    /**
+     * Returns true if a {@code Module} with the specified {@code Code} exists in the degree plan.
+     */
+    public boolean existingPlannerModules(Code plannerCode) {
```
Boolean methods should be named to sound like booleans.

Creastery commented on src/main/java/seedu/address/model/planner/Semester.java:
```diff
@@ -10,7 +10,8 @@
 
     public static final String MESSAGE_SEMESTER_CONSTRAINTS =
-            "Semester should only be given in a number range from 1 to 4, S (e.g. 1) format and in valid form.";
+            "Semester should only contain numbers between 1 and 4. "
```
Since the range is small, we should define the valid range clearly:
`either 1, 2, 3 or 4`

Creastery commented on src/main/java/seedu/address/model/planner/Year.java:
```diff
@@ -10,7 +10,8 @@
 
     public static final String MESSAGE_YEAR_CONSTRAINTS =
-            "Year in an university should only be given in Y (e.g. 1) format and in valid form.";
+            "Year should only contain numbers between 1 and 4. "
```
See above comment.

Creastery commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -290,6 +310,20 @@ public void removeDegreePlanner(DegreePlanner key) {
+    /**
+     * Returns a degree planner whose year and semester match the input year and semester.
+     */
+    public DegreePlanner getDegreePlanner(Year year, Semester semester) {
```
This creates unnecessary coupling between `DegreePlanner` with `Year` and `Semester`.
Since you are only using this method once in `PlannerAddCommand`, there is no need to extract this to an method of `AddressBook`.

WendyBaiYunwei replied:
I tried for a long time but still do not know how to get the degree planner list without having a method in Model. (once the method is in model, it also needs to be in addressbook if i'm not wrong)

To remove the coupling, I modified the implementation by reusing a method in addressbook, getDegreePlannerList. Hope it is acceptable.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,92 @@
+            + PREFIX_CODE + "CODE "
+            + "[" + PREFIX_CODE + "CODE]... "
+            + PREFIX_YEAR + "YEAR "
+            + PREFIX_SEMESTER + "SEMESTER\n"
```
Let's make it easier for users to understand how to use the application by changing:

```
planner_add year/YEAR sem/SEMESTER code/CODE [code/CODE]...
```

to:

```
planner_add code/CODE [code/CODE]... year/YEAR sem/SEMESTER 
```

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,92 @@
+
+    public static final String COMMAND_WORD = "planner_add";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds one/multiple modules to the degree plan. "
```
See https://github.com/CS2113-AY1819S2-T09-1/main/pull/128#discussion_r270641397

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,92 @@
+            + PREFIX_CODE + "CS2113T "
+            + PREFIX_CODE + "CS2100 "
+            + PREFIX_YEAR + "2 "
+            + PREFIX_SEMESTER + "2 ";
```
See above comment.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,92 @@
+    public static final String MESSAGE_MODULE_DOES_NOT_EXIST = "Some/one of the modules do"
+            + " not exist in the module list";
+    private Year yearToAdd;
+    private Semester semToAdd;
```
You are not actually adding a new `Year` or `Semester` when trying to add module codes to a `DegreePlanner` with the corresponding `Year` and `Semester`. 

This should be named as `year` and `semester` respectively.

WendyBaiYunwei replied:
I changed to yearToAddTo and semesterToAddTo to avoid repetition with the parsed arguments (also year and semester). Though they are not perfect names (a bit complicated), I believe that the names are accurate in meaning now.

WendyBaiYunwei replied:
and should be acceptabe?

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,92 @@
+    private Set<Code> codesToAdd;
+
+    /**
+     * Creates a PlannerAddCommand to add the specified {@Code code} to the degree plan.
```
`{@Code code}` -> `{@code code(s)}`

WendyBaiYunwei committed:
planner_add: change boolean method name

WendyBaiYunwei committed:
planner_add: modify year and sem messages

WendyBaiYunwei committed:
planner_add: remove unnecessary coupling
there is unnecessary coupling between DegreePlanner and Year and Semester.
let's remove it by reusing method getDegreePlannerList in Addressbook instead

WendyBaiYunwei committed:
planner_add: enhance code
there is unnecessarily complicated code that affects performance.
let's enhance and simplify it.

WendyBaiYunwei committed:
planner_add: modify messages

WendyBaiYunwei commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,94 @@
+            + PREFIX_CODE + "CS2100";
+
+    public static final String MESSAGE_SUCCESS = "The new module(s) added to degree plan: %1$s";
+    public static final String MESSAGE_DUPLICATE_MODULE = "Some/one of the module(s) already exist in degree plan";
```
I was hesitating whether to use modules or module(s) in this line. As I am not sure whether module(s) is grammatically acceptable. However, modules can be inaccurate as well. Perhaps we can check again with tutors on this afterwards?

Creastery replied:
`module(s)` is a shorthand term to signal other word endings (i.e. `module` and `modules`), and is both singular and plural. The use of such shorthand term is widely acceptable.

The problem is not the use of shorthand term, but rather, is how it is being used.

```
The new module(s) added to degree plan
```
can be rephrased to:
```
Added new module(s) to the year ... semester ... of the degree plan: ...
```

Similarly,
```
"Some/one of the module(s) already exist in degree plan"
```
can be rephrased to:
```
"The module (...) already exist in the degree plan"
```

Creastery replied:
`module(s)` is a shorthand term to signal other word endings (i.e. `module` and `modules`), and is both singular and plural. The use of such shorthand term is widely acceptable.

The problem is not the use of shorthand term, but rather, is how it is being used.

```
The new module(s) added to degree plan
```
can be rephrased to:
```
Added new module(s) to the year ... semester ... of the degree plan: ...
```

Similarly,
```
"Some/one of the module(s) already exist in degree plan"
```
can be rephrased to:
```
"The module (...) already exist in the degree plan"
```

WendyBaiYunwei replied:
I mean `module(s)` is acceptable, but I do not know whether `one of the module(s)` is acceptable because `one of` should be followed by plural.

Creastery replied:
The correct usage of this shorthand term is:
```
One (some) of the module(s)
```

But that's besides the point. For the second message, since there can be more than one modules here, you should inform the user what module already exists in the degree plan.

So I was thinking, if you only print the first code that triggers the error, it could be:
```
The module (AA11111) already exist in the degree plan.
```

Or if you are printing all erroneous codes:
```
The module(s) (AA11111) already exist in the degree plan.
The module(s) (AA11111, BB2222) already exist in the degree plan
```

WendyBaiYunwei committed:
planner_add: minor enhancements (+2 squashed commit)

Squashed commit:

[d8f416a9] planner_add: fix AddCommandTest

[5729ddf4] planner_add: remove extra imports

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,94 @@
+    public static final String MESSAGE_SUCCESS = "The new module(s) added to degree plan: %1$s";
+    public static final String MESSAGE_DUPLICATE_MODULE = "Some/one of the module(s) already exist in degree plan";
+    public static final String MESSAGE_MODULE_DOES_NOT_EXIST = "Some/one of the module(s) do"
+            + " not exist in the module list";
```
See above comment.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,94 @@
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
+        requireNonNull(model);
+
+        DegreePlanner currentDegreePlanner = model.getDegreePlannerList().stream()
```
`currentDegreePlanner` -> `selectedDegreePlanner`.

Creastery commented on src/main/java/seedu/address/model/AddressBook.java:
```diff
@@ -271,6 +271,15 @@ public void addDegreePlanner(DegreePlanner degreePlanner) {
+    /**
+     * Returns true if a {@code Module} with the specified {@code Code} exists in the degree plan.
+     */
+    public boolean hasPlannerModule(Code plannerCode) {
```
Here you are trying to check if a code exists in any `DegreePlanner` within `UniqueDegreePlanner`.
This method should be named `hasPlannerCode` or even `containsCodeInPlanners` to be more precise.

Creastery commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -302,6 +308,11 @@ public boolean hasDegreePlanner(DegreePlanner planner) {
+    @Override
+    public ObservableList<DegreePlanner> getDegreePlannerList() {
+        return versionedAddressBook.getDegreePlannerList();
+    }
```
This method is unnecessary.
You should be using `getAddressBook().getDegreePlannerList()` from `PlannerAddCommand` instead.

Creastery commented on src/main/java/seedu/address/model/planner/Semester.java:
```diff
@@ -10,7 +10,8 @@
     public static final String MESSAGE_SEMESTER_CONSTRAINTS =
-            "Semester should only be given in a number range from 1 to 4, S (e.g. 1) format and in valid form.";
+            "Semester should only be either 1, 2, 3 or 4. "
+                    + "Semester should not be blank";
```
This can be shorten to a single line.

Creastery commented on src/main/java/seedu/address/model/planner/Year.java:
```diff
@@ -10,7 +10,8 @@
     public static final String MESSAGE_YEAR_CONSTRAINTS =
-            "Year in an university should only be given in Y (e.g. 1) format and in valid form.";
+            "Year should only be either 1, 2, 3 or 4. "
+                    + "Year should not be blank";
```
This can be shorten to a single line.

WendyBaiYunwei committed:
planner_add: change implementation

WendyBaiYunwei committed:
planner_add: update with upstream

WendyBaiYunwei committed:
planner_add: remove repeated parts

WendyBaiYunwei committed:
planner_add: implement checking of duplicate and absent modules

WendyBaiYunwei committed:
planner_add: remove hasPlannerCode method (+1 squashed commits)

Squashed commits:

[3e3bbbb2] planner_add: remove extra part

WendyBaiYunwei commented:
In addition to the requests, I also removed hasPlannerCode method to make the code more concise.

WendyBaiYunwei committed:
planner_add: adjust spacing

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,101 @@
+                .anyMatch(codes -> codes.contains(code))).collect(Collectors.toSet());
+
+        Set<Code> absentModuleCodes = codesToAdd.stream().filter(code -> !model.hasModuleCode(code))
+                .collect(Collectors.toSet());
```
This should be declared below the first if statement.

Creastery replied:
Also, this should be renamed to `nonExistentModuleCodes`.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,101 @@
+    private Set<Code> codesToAdd;
+
+    /**
+     * Creates a PlannerAddCommand to add the specified {@Code code(s)} to the degree plan.
```
This should be `{@code codes}`

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -0,0 +1,101 @@
+                .anyMatch(codes -> codes.contains(code))).collect(Collectors.toSet());
+
+        Set<Code> absentModuleCodes = codesToAdd.stream().filter(code -> !model.hasModuleCode(code))
+                .collect(Collectors.toSet());
```
This should be renamed to `nonExistentModuleCodes`.
Also, please move this below the next if statement.

WendyBaiYunwei committed:
planner_add: make changes according to review

WendyBaiYunwei committed:
planner_add: change spacing

Creastery reviewed (changes requested):
You didn't fix the unresolved issue. Please check again.

WendyBaiYunwei commented:
> You didn't fix the unresolved issue. Please check again.

I thought code(s) was acceptable according to https://github.com/CS2113-AY1819S2-T09-1/main/pull/109#discussion_r270642695.

I will change still.

WendyBaiYunwei committed:
planner_add: change expression

Creastery commented:
> > You didn't fix the unresolved issue. Please check again.
> 
> I thought code(s) was acceptable according to [#109 (comment)](https://github.com/CS2113-AY1819S2-T09-1/main/pull/109#discussion_r270642695).
> 
> I will change still.

That comment was to let you know that you should be using `@code`.

Please see https://github.com/CS2113-AY1819S2-T09-1/main/pull/109#discussion_r270770533

WendyBaiYunwei committed:
planner_add: change code to codes
